Initial Code Review

ewanek1 opened the pull request:
Hello! PR for code review 🙂. Thank you!

ewanek1 committed:
Add files via upload

WilliamBergamin commented on app.js:
```diff
@@ -0,0 +1,41 @@
+  signingSecret: process.env.SLACK_SIGNING_SECRET,
+  socketMode: true,
+  appToken: process.env.SLACK_APP_TOKEN,
+  //logLevel: LogLevel. 
```
Could this be some code that was intended to be deleted?

ewanek1 replied:
Ah yes! Thanks for catching that 🫡

WilliamBergamin commented on utils/timeZoneMap.js:
```diff
@@ -0,0 +1,71 @@
+const tzMap = {
```
This makes sense but I would like to know why you opted to place this in `timeZoneMap.js` rather then `constants.js`?

ewanek1 replied:
Hm I believe since the map was a more crucial part of my code I just separated it out in its own file.

WilliamBergamin commented on commands/convertTimeLogic.js:
```diff
@@ -0,0 +1,73 @@
+/* The `convertTimeCommandText` function parses 
+ * and converts a time between time zones 
+ */
+function convertTimeCommandText(text) {
```
What do you think about adding proper [JSdocs](https://jsdoc.app/) here to add information on what `text` is expected to be and what the outputs of the function is?

As an outsider it can sometimes be hard to determine this just from the code

WilliamBergamin commented on commands/convertTimeLogic.js:
```diff
@@ -0,0 +1,73 @@
+    return {
+      error: `Sorry, I don't recognize the time zone abbreviation ${fromZoneAbbr}`,
+    };
+  } else if (!toZone) {
```
Because the `if` statement returns early I don't think we need the `else` keyword, I think we could get away with just `if` blocks 🤔 what dod you think?

ewanek1 replied:
Yes I agree! I suppose another technique would be to store the error messages inside an array especially if I want to notify a user if both their to and from zones are invalid.

WilliamBergamin commented on commands/teamClock.js:
```diff
@@ -0,0 +1,88 @@
+      output_message = "Invalid input. Please use `/teamclock` or `/teamclock @usergroup`";
+    }
+  } catch (error) {
+    console.error("Error fetching user list:", error);
```
This might be overkill for the project but this function could accept some sort of `context` input parameter that contains a `logger`, this way your entire project could use the Bolt logger 💡

zimeg replied:
Edit: Heh I realize now I'm making similar comments elsewhere - using the `context` value is a very interesting idea though! 🧠 ✨

WilliamBergamin commented on tests/convertLogic.test.js:
```diff
@@ -0,0 +1,105 @@
+const { convertTimeCommandText } = require("../commands/convertTimeLogic.js");
```
Nice 💯

WilliamBergamin commented on workflows/conversionWorkflows.js:
```diff
@@ -0,0 +1,99 @@
+/* Formats a moment object into a time string and its timezone abbreviation
+ * Example: 3pm EST
+ */
+function formatTimeWithAbbr(momentObject) {
```
Sweet 💯

WilliamBergamin commented on app.js:
```diff
@@ -0,0 +1,41 @@
+// Initializing Bolt app with Socket Mode
+const app = new App({
+  token: process.env.SLACK_BOT_TOKEN,
+  signingSecret: process.env.SLACK_SIGNING_SECRET,
```
💭 Socket mode apps do not need the `signing secret` it is used to validate HTTP requests coming from Slack

ewanek1 replied:
Ohh thanks for that!

WilliamBergamin commented on README.md:
```diff
@@ -1 +1,89 @@
+  Example: `/convert 3pm est pst`  
+  Converts a time between two time zones. Inputs are not case sensitive.
+
+- **`/time [time] @user`**  
```
Should this be `/timefor` instead of `/time`?

Does this command only accept one parameter?

ewanek1 replied:
Yes thanks for catching this. I'll fix it. I initially wrote this as /time <some time> @user. But I thought about it some more and figured a person would use their current time.

WilliamBergamin commented on commands/timeFor.js:
```diff
@@ -0,0 +1,76 @@
+  if (parts.length !== 1) {
+    return {
+      error:
+        "Invalid input format. Please use the `/time_for @user`\nExample: `/time_for @username`",
```
I think there might be a typo in the error message 🤔 how do you think we could improve this so that we maintain consistency across the project?

ewanek1 replied:
Got it - timefor not time_for.

WilliamBergamin reviewed (commented):
Overall this is great work 💯 its an awesome example of how to use the features of Bolt 🚀  

Some small things to consider:

When opening PRs, its helpful for reviewers to include things like a testing steps, Demo video/images, a short description of what the changes are etc. For this PR I had a decent amount of context on the app and could follow the README, but it isn't always the case

Including a [.gitignore](https://git-scm.com/docs/gitignore) file in your project is useful in order to filter out files that shouldn't be checked into the repositories, for example the `node_modules` folder or the `.env` file that contains the values for the environment variables.

Linters and formatters are great tools to facilitate collaboration on projects, they force every maintainer to adhere to a set of rules. In our [SDKs](https://github.com/slackapi/bolt-js/blob/5baeaeea5069b43bef3280b17416c323712682c0/package.json#L31C5-L32C75) we use [biome](https://github.com/biomejs/biome)

You've written some great unit tests, leveraging tools like [c8](https://www.npmjs.com/package/c8) to figure out exactly what sections of your code are getting covered is extremely useful. Trying to reach 100% code coverage is overkill, but these types of tools help highlight "critical" sections that may have been missed
 
I don't think this is needed for your project but using some sort of continuous integration pipeline to run the linters, unit tests and code coverage on every change introduced in a project can greatly improve the reviewers/maintainers confidence in the change. It helps to ensure that new changes do not create unexpected behaviors. In out [SDKs we use github actions](https://github.com/slackapi/bolt-js/tree/main/.github/workflows) to do this.

zimeg commented on commands/convertTime.js:
```diff
@@ -0,0 +1,17 @@
+    return;
+  }
+
+  await respond({ text: result });
```
📝 I'm a big fan of treating errors as values, but this might miss an unexpected `throw` from calling code, or even from a failing `ack` above!

Some examples in documentation might not show this at the moment, but a nice pattern to use can be found in this sample:

https://github.com/slack-samples/bolt-js-starter-template/blob/main/listeners/commands/sample-command.js

Changing the returning `error` to a [`throw`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw) might take additional logic to parse and be excessive, but I do think a `try`/`catch` at the listener level is most useful to avoid a crashed program 🤓

zimeg commented on commands/teamClockLogic.js:
```diff
@@ -0,0 +1,66 @@
+ * Expects input: '/teamclock @usergroup' for usergroup
+ */
+
+const moment = require("moment-timezone");
```
⏳ The time happenings included can no doubt use functions from elsewhere and I'm wondering if the [`Intl.DateTimeFormat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat) class was considered instead of `moment-timezone`?

AFAICT it might be a suitable replacement, albeit with not as concise syntax... But I do tend to prefer fewer dependencies when possible!

zimeg commented on commands/teamClockLogic.js:
```diff
@@ -0,0 +1,66 @@
+  { start: "16:00", end: "17:00", message: "Getting late 🕓" },
+  { start: "17:00", end: "20:00", message: "After hours 🌙" },
+  { start: "20:00", end: "24:00", message: "After hours 🌙" },
+  { start: "00:00", end: "09:00", message: "Sleeping 😴" },
```
🤔 Some of these time ranges have overlap at the `end` time which was confusing to me at a glance.

I'm somewhat curious if an ordered `status` list could be used instead to avoid some of the special case boundary checking below?

```js
const timeRanges = [
   { start: "00:00", message: "Sleeping 😴" },
   { start: "09:00", message: "Available ✅" },
   { start: "16:00", message: "Getting late 🕓" },
   { start: "17:00", message: "After hours 🌙" },
];
```

A following check can iterate in reverse here (or somehow else) to find the closest `start` time here for the corresponding message, though I admit that's makes the ordering important.

zimeg commented on commands/teamClockLogic.js:
```diff
@@ -0,0 +1,66 @@
+      const status = teamClockCommandText(curTime);
+
+      // Format output message line 
+      messages.push(`${user.real_name}: ${curTimeFormatted}  ⇔  ${status}`);
```
👁️‍🗨️ IIRC the `real_name` value isn't guaranteed here - formatting this to prefer the `<@${user.id}>` might be most safe for all outputs!

ewanek1 replied:
Thank you 😊

zimeg commented on commands/timeFor.js:
```diff
@@ -0,0 +1,76 @@
+const { WebClient } = require("@slack/web-api");
+const { convertTimeBetweenUsers } = require("./timeForLogic.js"); 
+const client = new WebClient(process.env.SLACK_BOT_TOKEN);
```
```suggestion
```

🪓 Passing the `app.client` to these functions might be a best practice to ensure the same client "options" are kept throughout the app.

I don't believe `@slack/bolt` changes these much, but if retry logic was changed during app initialization we'd want to make sure it matches here too!

zimeg commented on app.js:
```diff
@@ -0,0 +1,41 @@
+// Start Bolt app
+(async () => {
+  await app.start();
+  console.log("⚡️ Bolt app is running!");
```
```suggestion
  app.logger.info("⚡️ Bolt app is running!");
```

🗣️ Small preference to use the included `logger` for outputs we want to keep after development. I believe `console` is alright otherwise, but doesn't allow the log level to be changed in later code.

📚 https://tools.slack.dev/bolt-js/concepts/logging

zimeg commented on commands/teamClock.js:
```diff
@@ -0,0 +1,88 @@
+  // Format the user availability messages using their time zones
+  const userMessages = formatUserAvailabilityMessages(usersDataForProcessing);
+
+  await respond(userMessages.join("\n"));
```
```suggestion
  const message = userMessages.join("\n");
  await respond(message);
```

🗣️ Personal preference to pass explicit variables as arguments. Some runtimes can evaluate this as a function argument to call **within** `respond` instead of as a string.

I believe `javascript` interprets this as a string, but it's not always clear to me...

ewanek1 replied:
Gotcha! 🤝

zimeg commented on commands/teamClock.js:
👁️‍🗨️ I like that logic for each command is separated, but the separation between `teamClock` and `teamClockLogic` isn't so clear to me and I might favor a single file for all of this?

The `listeners` pattern in [samples](https://github.com/slack-samples/bolt-js-starter-template/tree/main/listeners) can be useful too once projects gain enough functionalities :sparkles:

zimeg commented on manifest.json:
```diff
@@ -0,0 +1,138 @@
+        "org_deploy_enabled": true,
+        "socket_mode_enabled": true,
+        "token_rotation_enabled": false,
+        "hermes_app_type": "remote",
```
```suggestion
```

🪓 I believe this is leftover but the same as `function_runtime`...

zimeg commented on package.json:
```diff
@@ -0,0 +1,19 @@
+  },
+  "dependencies": {
+    "@slack/bolt": "^4.4.0",
+    "@slack/web-api": "^7.9.3",
```
```suggestion
```

We might have access to the `client` from `@slack/bolt`!

zimeg commented on utils/constants.js:
🗣️ IMHO `utils` often hints at the need for another directory!

The time details that follow, and perhaps timezone specific logic for various listeners, could be moved into paths like so?

```sh
app.js
- listeners
  - commands
  - functions # for workflow steps
- tests
- timezones
```

The exact logic splits might not be right, but I do think similar organization can make the separation of Slack logic and timezone handling more clear 👾

ewanek1 replied:
Mmm I see. Thank you!

zimeg commented on workflows/conversionWorkflows.js:
```diff
@@ -0,0 +1,99 @@
+}
+
+// Export functions
+module.exports = {
+  handleTimeZoneConversionWF,
+  handleUserConversionWF,
+};
```
🤖 Exporting these functions into separate files might make future additions more obvious since fewer questions of where something should go might appear.

The shared logic within complicates that a smidge, but I think for app setups it might still be preferred.

zimeg reviewed (commented):
@ewanek1 This is an impressive setup all around!

I left a handful of comments building off of what @WilliamBergamin shared, but I do want to callout a few thoughts I was having:

- Reducing dependencies: I'm not sure if the native date handling of `javascript` can substitute the "moment" package, but for ongoing maintenance that might be optimal if so! Also, we might want to use `@slack/bolt` as the interface for all Slack interactions in this app.
- Separating handlers: Listeners and logic and handlers alike are found in similar places which is nice for fast context but can cause confusion in later enhancements. Some ideas on changes to this were left in comment!
- Error handling: Generous use of `try`/`catch` would move this app towards production readiness! Logging error cases might follow that, but I'm hoping most to avoid unexpected crashes.

Overall though, this is very solid development with a great grasp on the event driven setups of `bolt`! 🤖

ewanek1 committed:
Update README.md

Fixes the typo in the `timefor` command and removes the necessity for a signing secret since Chrono uses socket mode.

ewanek1 committed:
Update app.js

Removes signing secret because Chrono uses socket mode.

Commit:
Addresses PR feedback from reviewers after resetting branch. 

- Removed timefor slash command typos.
- Removed signing secret mentionings since socket mode is used.
- Removed the timeZoneMap.js file and moved contents into constants.js for better organization.
- Removed the 'else' in the if/else statement in convertTimeLogic.js.
- Addressed the overlap in the timeRanges array.
- Handled cases where real_name is not guaranteed in teamClockLogic.js.

WilliamBergamin reviewed (approved):
Great app 💯 most comments seem to be address 🚀
